src/main/scala/com/cloudera/sparkts/DateTimeIndex.scala

```diff
@@ -261,14 +359,280 @@ class IrregularDateTimeIndex(
   }
 
   override def equals(other: Any): Boolean = {
-    val otherIndex = other.asInstanceOf[IrregularDateTimeIndex]
-    otherIndex.instants.sameElements(instants)
+    other match {
+      case otherIndex: IrregularDateTimeIndex =>
+        otherIndex.instants.sameElements(instants)
+      case _ => false
+    }
   }
 
   override def toString: String = {
     "irregular," + dateTimeZone.toString + "," +
       instants.map(longToZonedDateTime(_, dateTimeZone).toString).mkString(",")
   }
+
+  override def millisIterator(): Iterator[Long] = {
+    new Iterator[Long] {
+      val instIter = instants.iterator
+
+      override def hasNext: Boolean = instIter.hasNext
+
+      override def next(): Long = instIter.next / 1000000L
+    }
+  }
+
+  override def zonedDateTimeIterator(): Iterator[ZonedDateTime] = {
+    new Iterator[ZonedDateTime] {
+      val instIter = instants.iterator
+
+      override def hasNext: Boolean = instIter.hasNext
+
+      override def next(): ZonedDateTime = longToZonedDateTime(instIter.next, dateTimeZone)
+    }
+  }
+
+  override def atZone(zone: ZoneId): IrregularDateTimeIndex = {
+    new IrregularDateTimeIndex(instants, zone)
+  }
+}
+
+/**
+ * An implementation of DateTimeIndex that holds a hybrid collection of DateTimeIndex
+ * implementations.
+ *
+ * Indices are assumed to be sorted and disjoint such that
+ * for any two consecutive indices i and j: i.last < j.first
+ *
+ */
+class HybridDateTimeIndex(
+    val indices: Array[DateTimeIndex],
+    val dateTimeZone: ZoneId = ZoneId.systemDefault())
+  extends DateTimeIndex {
+
+  private val sizeOnLeft: Array[Int] = {
+    var sum: Int = 0
+    (Array(0) ++ indices.init.map(_.size)).map {a =>
+      sum += a
+      sum
+    }
+  }
+
+  override def slice(interval: Interval): HybridDateTimeIndex = {
+    slice(ZonedDateTime.ofInstant(interval.getStart, dateTimeZone),
+      ZonedDateTime.ofInstant(interval.getEnd, dateTimeZone))
+  }
+
+  override def slice(start: ZonedDateTime, end: ZonedDateTime): HybridDateTimeIndex = {
```

Comment:
Does this need to return a HybridDateTimeIndex? Could just return a DateTimeIndex?